driver/src/main/java/org/neo4j/driver/internal/connector/socket/SSLSocketChannel.java

```diff
@@ -0,0 +1,518 @@
+     * {@code p = min(n, m)}.
+     * After the method call, the new position of {@code from.pos} will be {@code from.pos + p}, and similarly,
+     * the new position of {@code to.pos} will be {@code to.pos + p}
+     *
+     * @param from
+     * @param to
+     * @return
+     */
+    private int bufferCopy( ByteBuffer from, ByteBuffer to )
+    {
+        if ( from == null || to == null )
+        {
+            return 0;
+        }
+
+        int i;
+        for ( i = 0; to.remaining() > 0 && from.remaining() > 0; i++ )
+        {
+            to.put( from.get() );
+        }
+        return i;
+    }
+
+    /**
+     * Create network buffers and application buffers
+     *
+     * @throws IOException
+     */
+    private void createBuffers() throws IOException
+    {
+        SSLSession session = sslEngine.getSession();
+        int applicationBufferSize = session.getApplicationBufferSize();
+        int networkBufferSize = session.getPacketBufferSize();
+        bufferMax = Math.max( applicationBufferSize * 2, networkBufferSize * 2 );
+
+        createBuffers( applicationBufferSize, networkBufferSize );
+    }
+
+    void createBuffers( int appBufferSize, int netBufferSize )
+    {
+        plainOut = ByteBuffer.allocateDirect( appBufferSize );
+        plainIn = ByteBuffer.allocateDirect( appBufferSize );
+        cipherOut = ByteBuffer.allocateDirect( netBufferSize );
+        cipherIn = ByteBuffer.allocateDirect( netBufferSize );
+    }
+
+    /**
+     * Create SSLEngine with the SSLContext just created.
+     *
+     * @param host
+     * @param port
+     */
+    private void createSSLEngine( String host, int port )
+    {
+        sslEngine = sslContext.createSSLEngine( host, port );
+        sslEngine.setUseClientMode( true );
+    }
+
+    /**
+     * Init SSLContext with key store and trust key store.
+     * <p>
+     * For one-way SSL where a cert is only held by the server, the SSL could only be established if the client
+     * trusts the server's cert. If the server cert is created by our own, the choices that we have to make the client
+     * trusting the server cert are listed as follows:
+     * <ul>
+     * <li>
+     * Have the server cert signed by a CA which is also trusted by the client.</li>
+     * <li>
+     * Have the server cert signed by a corporate CA and make sure that the client also trust it.</li>
+     * <li>
+     * Add the server cert to the client's trusted cert list.</li>
+     * </ul><p>
+     * For the 2nd or 3rd option, what we could do is to add the cert to the trusted store used by the client. The
+     * trusted store could be created by {@code keytool} with the server cert. The trusted store could be loaded here
+     * or be set with {@code javax.net.ssl.trustStore} system property.
+     *
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    private void initSSLContext() throws GeneralSecurityException, IOException
```

Comment:
@zhenlineo I am probably reading this wrong, but it looks like you initialise the context to never have keys, and to never bother with trust. Does that get configured later somehow? It there a test that proves users _can_ use keys and trust chains?